observer.go

```diff
@@ -0,0 +1,65 @@
+package opentracing
+
+// Observer can be registered with the Tracer to recieve notifications
+// about new Spans.
+type Observer interface {
+	OnStartSpan(sp Span, operationName string, options StartSpanOptions) SpanObserver
+}
+
+// SpanObserver is created by the Observer and receives notifications
+// about other Span events.
+type SpanObserver interface {
+	OnSetOperationName(operationName string)
+	OnSetTag(key string, value interface{})
+	OnFinish(options FinishOptions)
+}
+
+// observer is a dispatcher to other observers
```

Comment:
I think this and below are implementation details that shouldn't be in the opentracing-go (especially since they are private and as such unusable anyway).

We need more people to look at the observer API before committing it. @bhs you mentioned a separate SPI, any thoughts?  At the moment this observer API is somewhat standalone, i.e. it can only be plugged into the tracer via vendor-specific initialization process (e.g. in Jaeger via tracer options).